src/instance/ffi_call.rs

```diff
@@ -0,0 +1,59 @@
+use std::os::raw::c_int;
+
+use crate::err::{error_on_minusone, SignedInteger};
+use crate::types::any::PyAnyMethods;
+use crate::{ffi, Py2, PyErr, PyResult, Python};
+
+/// Internal helper to convert raw ffi call results such as pointers
+/// or integers into safe wrappers.
+///
+/// `unsafe` to implement because it is highly likely this trait is
+/// passed a pointer, and is free to do interpret it as it sees fit.
+pub(crate) unsafe trait FromFfiCallResult<'py, RawResult>: Sized {
+    fn from_ffi_call_result(py: Python<'py>, raw: RawResult) -> PyResult<Self>;
```

Comment:
Staring at this again I wonder if it should have been a safe trait with the method being `unsafe fn from_ffi_call_result`. Both are a bit awkward because only the implementation for `*mut ffi::PyObject` actually needs to be `unsafe`. Anyway, this is only an internal API so it doesn't need to be perfect...

Reply:
Reading the comment above, why does it have to be unsafe at all? "May get a pointer and do unsafe stuff with it" doesn't sound like a strong argument :)

Reply:
I'm generally of the understanding that the pointer alone can't communicate that's it's safe to read.

Probably the right approach is to combine this with some wrapper along the lines of #1809 so that the trait doesn't need to be unsafe at all. I'll keep playing.